src/engine/openmp.cc

```diff
@@ -83,10 +83,10 @@ void OpenMP::set_reserve_cores(int cores) {
 
 int OpenMP::GetRecommendedOMPThreadCount(bool exclude_reserved) const {
 #ifdef _OPENMP
-  if (omp_num_threads_set_in_environment_) {
-    return omp_get_max_threads();
-  }
   if (enabled_) {
+    if (omp_num_threads_set_in_environment_) {
+      return omp_get_max_threads();
+    }
```

Comment:
this is the correct behavior, but i wonder if this will cause performance issues, for certain scenarios where clients of this class depend on GetRecommendedOMPThreadCount() to return the right number even if set_enabled is false. I only see set_enabled used in initialize so i think it should be fine. Other (probably less riskier) option is to set omp_num_threads_set_in_environment_ to false in the child process, but since it is private it will require additional setter and getter.

Reply:
I don't see it as a big risk, I went through the usage of the method. Maybe I missed something but the only non omp usage seems this one according to my IDE.

https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_utils.cc#L38

Reply:
okay i guess its fine. only other thing that worries me is the pragma omp parallel for used without any num_threads. I was thinking most should use GetRecommendedThreadCount, but there are a lot of places in code which use omp parallel for without a thread count. You mentioned that you called set_omp_num_threads in the child handler of pthread_atfork. didnt that help ?

Reply:
looks that's implementation dependent. We should test more, but I think let's get this done first.